### PR TITLE
Inter-domain loss using Prototype KL divergence

### PR DESCRIPTION
**Pull request by adv010**
POC pipeline created.. Testing underway

### CODE REVIEW

File: pcdet/models/roi_heads/pvrcnn_head.py [hunk 1]
```diff
@@ -159,6 +170,18 @@ def forward(self, batch_dict, disable_gt_roi_when_pseudo_labeling=False):
         batch_size_rcnn = pooled_features.shape[0]
         pooled_features = pooled_features.permute(0, 2, 1).\
             contiguous().view(batch_size_rcnn, -1, grid_size, grid_size, grid_size)  # (BxN, C, 6, 6, 6)
+        
+        batch_dict['pooled_features'] =  pooled_features.view(batch_dict['batch_size'],batch_dict['roi_labels'].shape[1],-1, grid_size, grid_size, grid_size)
```
Comment (shashankag14) [inline pcdet/models/roi_heads/pvrcnn_head.py:167]: For better readability, I think we can reshape of `pooled_features` from (BxN, C, 6, 6, 6) to (B, N, C, 6, 6, 6) inside `calc_prototype()` instead of doing it at this stage.

File: pcdet/models/roi_heads/pvrcnn_head.py [hunk 2]
```diff
@@ -159,6 +170,18 @@ def forward(self, batch_dict, disable_gt_roi_when_pseudo_labeling=False):
...
+        batch_dict['pooled_features_lbl'] = batch_dict['pooled_features'][batch_dict['labeled_inds']]
```
Comment (shashankag14) [inline pcdet/models/roi_heads/pvrcnn_head.py:168]: I think we can remove L175,176 as the masks can be created inside `pooled_features` itself.

File: pcdet/models/roi_heads/pvrcnn_head.py [hunk 3]
```diff
@@ -178,3 +201,37 @@ def forward(self, batch_dict, disable_gt_roi_when_pseudo_labeling=False):
...
 
         return batch_dict
+
+#TODO - keep source meaned to classwise, unlabeled non-mean. Ablation  mean
+    def calc_prototype(self,batch_dict,enableViewA=False):
+        if enableViewA: # weakly augmented through Student
+            src_prototype = self.src_prototypeViewA
+            tar_prototype = self.target_prototypeViewA
+        else: # strongly augmented  through Student
+            src_prototype = self.src_prototype
```
Comment (shashankag14) [inline pcdet/models/roi_heads/pvrcnn_head.py]: Source prototype (reference prototype) needs to be created only once in each iteration. We can remove the creation of prototype for strongly augmented data, i.e. `src_prototype`

File: pcdet/models/roi_heads/pvrcnn_head.py [hunk 4]
```diff
@@ -178,3 +201,37 @@ def forward(self, batch_dict, disable_gt_roi_when_pseudo_labeling=False):
...
+            tar_prototype = self.target_prototype
+        # Generate Source classwise prototype ()
+        for i in range(1, len(self.class_dict)+1):
+            cls_mask = batch_dict['roi_labels'][batch_dict['labeled_inds']] == i
+            key = self.class_dict[i]
+            cur_proto = (batch_dict['pooled_features'][batch_dict['labeled_inds']][cls_mask]).mean(dim=0)
+            if batch_dict['cur_iteration']< self.start_iter:
```
Comment (shashankag14) [inline pcdet/models/roi_heads/pvrcnn_head.py]: Maybe another flag (eg. `SSL_MODE`) can be created to set `self.model_config.START_ITER` in the config to 0 if running the model in SSL setting, i.e. with pre-trained model. 

Else it should take the value as mentioned in the config file for `self.model_config.START_ITER` because then it would be running the model in SSDA i.e. without pre-trained model. 

For eg.
```python
self.start_iter = 0 if self.model_config.SSL_MODE else self.model_config.START_ITER
``` 
`self.start_iter` can be initialized with something as given above at the beginning in `pv_rcnn_ssl.py`

File: pcdet/models/roi_heads/roi_head_template.py [hunk 5]
```diff
@@ -305,3 +451,16 @@ def generate_predicted_boxes(self, batch_size, rois, cls_preds, box_preds):
         batch_box_preds[:, 0:3] += roi_xyz
         batch_box_preds = batch_box_preds.view(batch_size, -1, code_size)
         return batch_cls_preds, batch_box_preds
+    
+    def get_proto_inter_loss(self):
```
Comment (shashankag14) [inline pcdet/models/roi_heads/roi_head_template.py]: This can be called inside `get_loss()` in `roi_head_template.py` where all other losses of RCNN are computed.

File: tools/cfgs/kitti_models/pv_rcnn_ssl_60.yaml [hunk 6]
```diff
@@ -186,6 +187,10 @@ MODEL:
             NSAMPLE: [16, 16]
             POOL_METHOD: max_pool
 
+        PROTOTYPE:
+            MOMENTUM : 0.9
+            START_ITER : 20
```
Comment (shashankag14) [inline tools/cfgs/kitti_models/pv_rcnn_ssl_60.yaml:194]: Is this 20 iterations or epochs?

File: pcdet/models/roi_heads/pvrcnn_head.py [hunk 7]
```diff
@@ -44,6 +44,14 @@ def __init__(self, input_channels, model_cfg, num_class=1,
         self.init_weights(weight_init='xavier')
 
         self.print_loss_when_eval = False
+        self.class_dict = {1:'Car', 2 :'Ped', 3:'Cyc'}
+        self.src_prototype = {'Car': None, 'Ped' : None, 'Cyc' : None}
```
Comment (shashankag14) [inline pcdet/models/roi_heads/pvrcnn_head.py]: This might need changes for distributed mode setting. Something like TorchMetrics.

File: pcdet/datasets/__init__.py [hunk 8]
```diff
@@ -72,7 +72,7 @@ def build_dataloader(dataset_cfg, class_names, batch_size, dist, root_path=None,
     dataloader = DataLoader(
         dataset, batch_size=batch_size, pin_memory=True, num_workers=workers,
         shuffle=(sampler is None) and training, collate_fn=dataset.collate_batch,
-        drop_last=False, sampler=sampler, timeout=0
+        drop_last=False, sampler=sampler, timeout=0, worker_init_fn=partial(common_utils.worker_init_fn, seed=seed)
```
Comment (shashankag14) [inline pcdet/datasets/__init__.py:75]: Why is this required?

File: pcdet/models/detectors/pv_rcnn_ssl.py [hunk 9]
```diff
@@ -1,49 +1,95 @@
...
 
     def forward(self, batch_dict):
         if self.training:
             labeled_mask = batch_dict['labeled_mask'].view(-1)
             labeled_inds = torch.nonzero(labeled_mask).squeeze(1).long()
+            batch_dict['labeled_inds'] = labeled_inds
             unlabeled_inds = torch.nonzero(1-labeled_mask).squeeze(1).long()
+            batch_dict['unlabeled_inds'] = unlabeled_inds
             batch_dict_ema = {}
+            batch_dict['store_scores_in_pkl'] = self.model_cfg.STORE_SCORES_IN_PKL
```
Comment (shashankag14) [inline pcdet/models/detectors/pv_rcnn_ssl.py:92]: Whats the use of this in this file?

File: pcdet/models/roi_heads/roi_head_template.py [hunk 10]
```diff
@@ -305,3 +451,16 @@ def generate_predicted_boxes(self, batch_size, rois, cls_preds, box_preds):
...
+        cls_map =  {0:'Car',1:'Ped',2:'Cyc'}
+        #TODO: self.src_prototype is a dict. calculate classwise CE
+        cos_viewA = []
+        cos_viewB = []
+        for i in range(3):
+            cos_viewA.append(F.cosine_similarity(self.src_prototypeViewA[cls_map[i]],self.target_prototypeViewA[cls_map[i]]))
+            cos_viewB.append(F.cosine_similarity(self.src_prototypeViewA[cls_map[i]],self.target_prototypeViewBA[cls_map[i]]))
+        cos_viewA = torch.stack(cos_viewA)
+        cos_viewB= torch.stack(cos_viewB)
+        loss =  F.kl_div(cos_viewA, cos_viewB)
```
Comment (shashankag14) [inline pcdet/models/roi_heads/roi_head_template.py]: Maybe it would be helpful to add a plot to monitor the loss over the itr/epochs. (This plot could be sent toTB logger)

File: pcdet/models/roi_heads/pvrcnn_head.py [hunk 11]
```diff
@@ -159,6 +171,20 @@ def forward(self, batch_dict, disable_gt_roi_when_pseudo_labeling=False):
...
             contiguous().view(batch_size_rcnn, -1, grid_size, grid_size, grid_size)  # (BxN, C, 6, 6, 6)
+        
+        batch_dict['pooled_features'] =  pooled_features.view(batch_dict['batch_size'],batch_dict['roi_labels'].shape[1],-1, grid_size, grid_size, grid_size)
+        batch_dict['pooled_features_lbl'] = batch_dict['pooled_features'][batch_dict['labeled_inds']]
+        batch_dict['pooled_features_ulb'] =  batch_dict['pooled_features'][batch_dict['unlabeled_inds']]
+
+
+        if not batch_dict['module_type'] == 'Teacher':
+            if batch_dict['module_type'] == 'Student': 
+                self.src_prototypeB,self.target_prototypeViewB = self.calc_prototype(batch_dict)
```
Comment (shashankag14) [inline pcdet/models/roi_heads/pvrcnn_head.py]: I think this code block (L180-187) needs to be refactored overall in terms of readability (eg. a simple if-elif-else statements to check the module_type instead of nested ones).

Also, we do not need `self.src_prototypeB` as a return object.

File: pcdet/models/roi_heads/pvrcnn_head.py [hunk 12]
```diff
@@ -178,3 +204,57 @@ def forward(self, batch_dict, disable_gt_roi_when_pseudo_labeling=False):
             self.forward_ret_dict = targets_dict
 
         return batch_dict
+
+#TODO - keep source meaned to classwise, unlabeled non-mean. Ablation  mean
+    def calc_prototype(self,batch_dict):
+        if batch_dict['module_type'] == "Student": # Strong augmentation through Student
```
Comment (shashankag14) [inline pcdet/models/roi_heads/pvrcnn_head.py]: Do we need L210-216? Can we avoid these steps and directly use the objects initialized in this class to update the prototypes?

File: pcdet/models/roi_heads/pvrcnn_head.py [hunk 13]
```diff
@@ -178,3 +204,57 @@ def forward(self, batch_dict, disable_gt_roi_when_pseudo_labeling=False):
...
+            cls_mask = batch_dict['roi_labels'][batch_dict['unlabeled_inds']] == i
+            key = self.class_dict[i]
+            cur_proto = (batch_dict['pooled_features'][batch_dict['unlabeled_inds']][cls_mask]).mean(dim=0)
+            if batch_dict['cur_iteration']< self.start_iter:
+               tar_prototype[key] = cur_proto
+            else :
+               tar_prototype[key] = self.momentum *tar_prototype[key] + (1 - self.momentum) * cur_proto   
+        return src_prototype,tar_prototype
+
+    def proto_WeakB(self,batch_dict):
```
Comment (shashankag14) [inline pcdet/models/roi_heads/pvrcnn_head.py]: (optional) Function name can be made better in terms of readability (eg. `create_viewB_prototype()`)

**Review by shashankag14 — COMMENTED** [review]
Please have a look at the review comments. 
In terms of the code flow, it seems fine, but please debug the main additions done to the code in terms of prototype consistency to make sure that the shapes of prototypes are as expected at different stages.

File: tools/cfgs/kitti_models/pv_rcnn_ssl_60.yaml (hunk 6 above)
Comment (adv010) [reply tools/cfgs/kitti_models/pv_rcnn_ssl_60.yaml:194]: Iterations,  I compare this value with batch_dict['cur_iterations']

File: pcdet/models/roi_heads/roi_head_template.py (hunk 10 above)
Comment (adv010) [reply pcdet/models/roi_heads/roi_head_template.py]: good idea! Adding to TODO

File: pcdet/models/roi_heads/pvrcnn_head.py (hunk 4 above)
Comment (adv010) [reply pcdet/models/roi_heads/pvrcnn_head.py]: I have [commited](https://github.com/adv010/OpenPCDet/commit/2c0812121d807c3c25717b2c9975c4fa96bb8474) a subsection named pretrained in yaml file to check for a pretrained model. 

I shall link this flag with the start_iter value as suggested

File: pcdet/models/roi_heads/pvrcnn_head.py (hunk 2 above)
Comment (adv010) [reply pcdet/models/roi_heads/pvrcnn_head.py:168]: I preferred the masking here for my readability. Instead of removing them, I have changed the indexing happening inside calc_prototype and made use of these pooled_features_lbl and pooled_features_ulb

File: pcdet/models/detectors/pv_rcnn_ssl.py (hunk 9 above)
Comment (adv010) [reply pcdet/models/detectors/pv_rcnn_ssl.py:92]: None as of now :) ... I might use it at some point though, unrequired as for now

File: pcdet/models/roi_heads/pvrcnn_head.py (hunk 13 above)
Comment (adv010) [reply pcdet/models/roi_heads/pvrcnn_head.py]: Agreed. Renamed to create_strong_rev_proto (rev to indicate the reverse-augmentation step )

File: pcdet/models/roi_heads/pvrcnn_head.py (hunk 4 above)
Comment (adv010) [reply pcdet/models/roi_heads/pvrcnn_head.py]: Implemented the ternary operator advised above with 1 small change -  set the start_iter as 1 if pretrained model present. Allowing  iter 0 (first iter) to just create prototypes using pretrained weights after loading.

**Review by shashankag14 — APPROVED** [review]
